PR for #4037: Added `Ctrl+z` Undo/Redo.

DeclineThyself opened the pull request:
This PR implements an undo history by saving copies of the chat every time `saveChat` is called.

 #4037
 `Ctrl` or `Meta` + `z` triggers undo.
 `Ctrl` or `Meta` + `Z` triggers redo.


```javascript
/**
 * Save a copy of chatData to chatHistory.
 * @param {ChatMessage[]} chatData
 */
export function saveChatSnapshot(chatData = chat){
    //Overwrite history that has been undone.
    chatHistory.splice(chatHistoryIndex + 1);

    chatHistory.push(structuredClone(chatData));
    chatHistoryIndex = chatHistory.length - 1;
}
```

This PR is incomplete, I'll finish it if @Cohee1207 approves of this implementation.

Known Issues:
- [X] There is no limit to the size of the `chatHistory` array.
- [X] ~~This may have a performance impact on large chats.~~ (0.2 seconds on a chat with 5k messages.)
- [X] This should be toggled in ~~`User Settings` > `Chat/Message Handling`.~~ Extension Menu.
- [ ] This does not work with [chatTree](https://github.com/SillyTavern/SillyTavern/pull/4573).
- [X] `loadChatSnapshot` should emit an event.
- [X] This should be refactored into a separate file.
- [X] Users may accidentally trigger an `Undo` by clicking `Ctrl + z` while a text input is not focused. 
- [X] This is not accessible to touch screens.
- [X] `saveChat` is often called when the chat has not been changed, this results in redundant undo entries. Instead, `saveChatSnapshot` should be called on edit, send and delete.

This could be an extension if `saveChat` emitted an event, but I'd prefer it to be part of ST.

## Checklist:

- [X] I have read the [Contribution guidelines](https://github.com/SillyTavern/SillyTavern/blob/release/CONTRIBUTING.md).

Commit:
Implemented `saveChatSnapshot` and `loadChatSnapshot`.

DeclineThyself commented:
/gemini summary

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/script.js:
```diff
@@ -372,6 +372,9 @@ export let name1 = default_user_name;
 /** @type {ChatMessage[]} */
 export let chat = [];
+/** @type {ChatMessage[][]} */
+export let chatHistory = [chat];
```
~~This allows for `redos` to an unsaved chat.~~
Nevermind, Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/32403f87ff872a7df133ba6809c92775fac993de

DeclineThyself commented on public/script.js:
```diff
@@ -6848,6 +6851,42 @@ export function saveChatDebounced() {
+
+    chatHistory.push(structuredClone(chatData));
+    chatHistoryIndex = chatHistory.length - 1;
+}
```
See https://github.com/SillyTavern/SillyTavern/pull/4819#issue-3665512618.

DeclineThyself commented on public/script.js:
```diff
@@ -6848,6 +6851,42 @@ export function saveChatDebounced() {
+        refreshSwipeButtons();
+
+        // Is this needed?
+        // eventSource.emit(event_types.CHAT_CHANGED, getCurrentChatId());
```
@Cohee1207, is Gemini correct?

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2562779627

DeclineThyself commented on public/script.js:
```diff
@@ -6848,6 +6851,42 @@ export function saveChatDebounced() {
+export async function loadChatSnapshot(index) {
+    if (chatHistory[index]) {
+        chatHistoryIndex = index;
+        chat = structuredClone(chatHistory[chatHistoryIndex]);
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/c5b0e056e64b30f0ab22eb43805297e473b8dea6

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2562779624

Cohee1207 commented:
Does this consider the memory footprint? I don't see any limits on how large the history may be. This has the possibility to exhaust the memory, fast.

DeclineThyself commented:
> Does this consider the memory footprint? I don't see any limits on how large the history may be. This has the possibility to exhaust the memory, fast.

No, not yet:

> This PR is incomplete, I'll finish it if @Cohee1207 approves of this implementation.
> Known Issues:
>     * There is no limit to the size of the `chatHistory` array.
>     * This may have a performance impact on large chats.
> ...

I'd appreciate guidance before I implement the rest.

Cohee1207 commented:
Before we proceed, let me ask the classic question: can this be implemented as an extension? What stops it besides a missing BEFORE_CHAT_SAVE event?

DeclineThyself commented:
> Before we proceed, let me ask the classic question: can this be implemented as an extension? What stops it besides a missing BEFORE_CHAT_SAVE event?

No, this could be an extension.
I can use the `edited/deleted/sent` events instead, and I'd also need to copy `isInputElementInFocus`.
But this feels like a simple easy to maintain feature that should be part of core.
And some extensions may need to handle an `Undo` or `Redo` event.

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Removed redundant `refreshSwipeButtons`.

Cohee1207 commented:
Let's move into an extension. If you don't want to make a downloadable one, start with a built-in.

Commit:
Moved `isInputElementInFocus` and `isModifiedKeyboardEvent` to `utils`.

Commit:
Moved `chatHistory` to extension.

DeclineThyself commented:
> Let's move into an extension. If you don't want to make a downloadable one, start with a built-in.

https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1de8c1ad37ea97e6431057cc585ab7f7332d4169

DeclineThyself commented:
@Cohee1207 

1. How's this so far https://github.com/SillyTavern/SillyTavern/commit/1de8c1ad37ea97e6431057cc585ab7f7332d4169?
2. For touch screens, would you prefer buttons or gestures? Where?
4. Should `chatHistory`'s max size setting be in `Chat/Message Handling` or `Extensions`?
5. Should I also add a max chat length setting?
6. https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2562797108
7. Should a warning popup appear the first time undo is clicked?
8. Which events should trigger `saveChatSnapshot`?
9. Should `undo`/`redo` save the chat?
10. Should an existing or new event be emitted by `loadChatSnapshot`?
11. Should `chatHistory` persist? e.g. `sessionStorage.setItem`.

Cohee1207 commented on public/scripts/extensions/ctrl-z/index.js:
```diff
@@ -0,0 +1,85 @@
+    }
+}
+
+function isInputElementInFocus() {
```
Import from utils?

DeclineThyself replied:
Thanks, I forgot.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/b6c848c606194e69ae5ad26f1f33ebee1bd08771

Cohee1207 commented on public/scripts/extensions/ctrl-z/index.js:
```diff
@@ -0,0 +1,85 @@
+
+        // Is this needed?
+        // await eventSource.emit(event_types.MESSAGE_DELETED, chat.length);
+        // eventSource.emit(event_types.CHAT_CHANGED, getCurrentChatId());
```
Not needed. CHAT_CHANGED is for the chat ID changes, not its contents. I think it needs to either emit DELETED or RECEIVED/SENT with related "rendered" events depending on whether the chat length increased or decreased. If unchanged - don't emit anything.

DeclineThyself replied:
Like this?: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/b2cf3460a5f4b1a4c7e505225cfc538c82cc5c6c

Cohee1207 commented on public/scripts/extensions/ctrl-z/index.js:
```diff
@@ -0,0 +1,85 @@
+eventSource.on(event_types.CHAT_CHANGED,  async () => await resetChatSnapshots(chat));
+
+//Snapshot the chat when a message is modified.
+const snapshotEvents = [event_types.MESSAGE_DELETED, event_types.MESSAGE_EDITED, event_types.MESSAGE_SENT]; //Incomplete list.
```
Should you also save on swiping, swipe deletion, etc.?

DeclineThyself replied:
How's this? I may have missed some.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/16701cc566496aa022f45c418d1d74f22a3340c0

Cohee1207 commented on public/scripts/extensions/ctrl-z/manifest.json:
```diff
@@ -0,0 +1,9 @@
+{
+    "display_name": "ctrl-z",
```
I think it needs a better name. From semantical perspective, "Chat History" won't fit as this term is reserved for a history of past chats. Something like "Chat Undo History" would fit

DeclineThyself replied:
Renamed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/fc42f9962571fa2db8e143901b16238e8ddef625

I'd prefer to leave `ctrl + z` in the `display_name` so users can understand at a glance.

Cohee1207 commented:
> How's this so far https://github.com/SillyTavern/SillyTavern/commit/1de8c1ad37ea97e6431057cc585ab7f7332d4169?

Added a couple of comments. The direction is good so far.

> For touch screens, would you prefer buttons or gestures? Where?

I think of putting two buttons into either hamburger or wand menus. Like this. 

<img height="300" alt="image" src="https://github.com/user-attachments/assets/45de7874-533d-473e-9fae-e9edd57d840a" />

> Should chatHistory's max size setting be in Chat/Message Handling or Extensions?

I think that would need a drawer in extensions menu similar to summary/TTS/whatever. So far I anticipate it having two toggles: enabled (off by default) and history size slider.

> Should I also add a max chat length setting?

Why? Also, if you worry about performance, why not storing diffs between states instead of full chat clones? That may require some research first on how to generate and apply diffs to arrays.

>  PR for #4037: Added Ctrl+z Undo/Redo. #4819 (comment)

Answered in a new review comment.

> Should a warning popup appear the first time undo is clicked?

What kind of warning? if it's explicitly enabled first, no warning is required.

> Which events should trigger saveChatSnapshot?

Anything that changes the chat state: message added by either role / edited / deleted / swiped.

> Should undo/redo save the chat?

Yes. At least using debounced save, all changes to chat are meant to be auto-saved since there's no explicit "save" operation except `/forcesave` command

> Should an existing or new event be emitted by loadChatSnapshot?

Answered in a review comment. Ideally it should reflect the type of change that was applied to the chat: edited a message, deleted, added, etc.

> Should chatHistory persist? e.g. sessionStorage.setItem.

I think no. Usually programs delete the editor history when you close them.

DeclineThyself commented:
Thank you!

I'll be working on this bug first: https://github.com/SillyTavern/SillyTavern/pull/4780#issuecomment-3586999202

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Imported `isInputElementInFocus`.

Commit:
Added `MESSAGE_SWIPE_STARTED` and `MESSAGE_SWIPE_ENDED` events.

Commit:
Debounced `saveChatSnapshot`.

Commit:
Emitted `MESSAGE_RECEIVED` and `MESSAGE_DELETED` events, added `CHAT_SNAPSHOT_LOADED` event.

Commit:
Save the chat after `loadChatSnapshot`.

Commit:
Save a snapshot on more events.

Commit:
Renamed to `ChatUndoHistory`.

Commit:
Added buttons to `options-content`.

Commit:
Added manual save buttons and settings.

DeclineThyself commented:
/gemini summary

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/scripts/utils.js:
```diff
@@ -2884,3 +2884,32 @@ export function createTimeout(ms, errorMessage = '') {
+        event.shiftKey ||
+        event.ctrlKey ||
+        event.altKey ||
+        event.metaKey);
```
Out of scope. I moved the function, I haven't changed it.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1711ec41f60526d462c647fc88fb5aeb0e274ef4

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,127 @@
+        return;
+    }
+
+    if (chatHistory[index]) {
```
No. This is intended.
I don't want to overwrite a chat that's longer than `maximumChatLength`.

Commit:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819#pullrequestreview-3517625506

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,127 @@
+        await eventSource.emit(event_types.CHAT_SNAPSHOT_LOADED, index);
+
+        if (newChat.length > chat.length) { await eventSource.emit(event_types.MESSAGE_RECEIVED); }
+        if (newChat.length < chat.length) { await eventSource.emit(event_types.MESSAGE_DELETED); }
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/718d28a3303b4fa168ffe673bd737ada954c6c82

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,127 @@
+
+
+    //Enforce the maximum chat length.
+    if (chat.length >= maximumChatLength) {
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/718d28a3303b4fa168ffe673bd737ada954c6c82

DeclineThyself commented:
/gemini review

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,128 @@
+        await eventSource.emit(event_types.CHAT_SNAPSHOT_LOADED, index);
+
+        if (newChat.length > oldChatLength) { await eventSource.emit(event_types.MESSAGE_RECEIVED); }
+        if (newChat.length < oldChatLength) { await eventSource.emit(event_types.MESSAGE_DELETED); }
```
@Cohee1207 I forgot about that.
Should, I trust `debounce` to handle it?

Cohee1207 replied:
I think the simplest would be "isRestoring" flag that temporarily holds making new snapshots.

DeclineThyself replied:
Is this okay? The events shouldn't conflict with anything.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1211c9e467ec8cb7db507aeb6da1708fda0de6fc

Cohee1207 replied:
Maybe, I would need to see how it behaves live.

Commit:
Fixed:
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2570614002
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2570614009

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,128 @@
+ * @param {number} index The chatHistory index to load.
+ */
+export async function loadChatSnapshot(index) {
+    const maximumChatLength = extension_settings[extensionName]?.max_length ?? 1000;
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/7b985bb388ebbcc90281e92eb2778311a2a9ceae

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,128 @@
+
+    //Enforce the maximum chat length.
+    if (chatData.length >= maximumChatLength) {
+        toast && toastr.error(t`It's in 'Extensions > Chat Undo History > Max chat length'`, t`You cannot save the chat because it's ${chat.length - maximumChatLength} messages longer than your max chat length limit (${maximumChatLength}). (Check Settings.)`);
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/7b985bb388ebbcc90281e92eb2778311a2a9ceae

DeclineThyself commented:
@Cohee1207 That's almost everything, The UI code took a while.

> Why? Also, if you worry about performance, why not storing diffs between states instead of full chat clones? That may require some research first on how to generate and apply diffs to arrays.

I'm worried about performance because this function runs a `StructuredClone` of chat on almost all `message` events. 
I'd like this to work with `feat/chat-tree`, so the diff will need to accept objects.
Would these two packages be acceptable?

- https://www.npmjs.com/package/deep-object-diff
- https://www.npmjs.com/package/@transformation-dev/deep-object-diff-apply

Context: https://github.com/mattphillips/deep-object-diff/issues/73#issuecomment-1668615197

> What kind of warning? if it's explicitly enabled first, no warning is required.

~~Should I leave the `ctrl-z` hotkey toggled off by default?~~ https://github.com/SillyTavern/SillyTavern/pull/4819/commits/4cc9b2be6ea677a7add048a941fd84466b663ce5

Commit:
Fixed redundant saves.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2571515880

Commit:
The undo hotkey is now toggled off by default.

DeclineThyself commented:
@Cohee1207, I had a few questions:
4819
~~https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2572236981~~
~~https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3588157446~~

4573
https://github.com/SillyTavern/SillyTavern/pull/4573#issuecomment-3586529605
~~https://github.com/SillyTavern/SillyTavern/pull/4573#discussion_r2572568618~~

Cohee1207 commented:
> Would these two packages be acceptable?

Ok.

Commit:
Added object diff packages.
https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3588157446

https://www.npmjs.com/package/deep-object-diff
https://www.npmjs.com/package/@transformation-dev/deep-object-diff-apply

Commit:
Refactored into ChatHistory class.

Commit:
Fixed hotkeys.

Commit:
ESLint.

Commit:
Only save chat diffs.

Commit:
Only save changed chats.

DeclineThyself commented:
/gemini review

DeclineThyself commented:
/gemini summary

Commit:
Swapped from max_history to max_chunks.

DeclineThyself commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,177 @@
+        }
+
+        //Enforce the maximum chat History length.
+        this.chatHistory.splice(0, this.chatHistory.length - maximumChatHistoryItems + 1);
```
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/ba134932e5385b8dbe77432d2faafc3cf197578d

Commit:
Fixed debounce_duration = 0.

Commit:
Added debug logging.

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Cohee1207 commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,189 @@
+    event_types.MESSAGE_RECEIVED,
+    event_types.MESSAGE_EDITED,
+    event_types.MESSAGE_DELETED,
+    event_types.MESSAGE_UPDATED,
```
MESSAGE_UPDATE follows MESSAGE_EDITED. Keep one

DeclineThyself replied:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/e11e83eb2c6900dcc6dac8703f02dc291bcf108d

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,189 @@
+    event_types.MESSAGE_EDITED,
+    event_types.MESSAGE_DELETED,
+    event_types.MESSAGE_UPDATED,
+    event_types.MESSAGE_FILE_EMBEDDED,
```
Should there be an event for deleted file and/or media file from message?

DeclineThyself replied:
Probably, But I wouldn't use it here. The media would remain deleted and only the message would be reverted.

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+export const chatHistory = new ChatHistory(chat);
```
Do you need to keep a reference to chat? It should be static actually and its reference never changes.

DeclineThyself replied:
I've written `ChatHistory` to be mostly compatible with the `chatTree`, so the reference will be necessary.

Cohee1207 commented on public/scripts/extensions.js:
```diff
@@ -211,6 +211,7 @@ export const extension_settings = {
         /** @type {string} */
         sort: 'dateAsc',
     },
+    ChatUndoHistory: {},
```
Let's keep the naming consistent and use `lowerCamel`. `snake_case` is discouraged for new extensions.

DeclineThyself replied:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/52617a6bf8d16e988befee8a8e8fe11763adc54a

Cohee1207 commented on public/scripts/extensions/undo/manifest.json:
To be consistent with the rest of built-in extensions, it should use `kebab-case` for folder naming
`chat-undo-history` would do.

DeclineThyself replied:
Can I use a consistent name for both the folder and extension?
If not, may I call it `undo`?

DeclineThyself replied:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/52617a6bf8d16e988befee8a8e8fe11763adc54a

Cohee1207 replied:
`undo` is fine for me

Cohee1207 commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,189 @@
+    event_types.MESSAGE_REASONING_DELETED,
+    event_types.MESSAGE_SWIPE_DELETED ];
+
+jQuery(async () => {
```
Feel free to replace this with a Immediately Invoked Function Expression
```js
(() => {
})();
```

DeclineThyself replied:
Should I update the other extensions too?
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/313e1aa00a1afe441146fbb6f173c657795a585b
I assumed `JQuery` was used to load after the DOM.

Cohee1207 replied:
DOM will be already loaded at this point where extension scripts are loaded. No need to change elsewhere.

Cohee1207 commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,189 @@
+     * Save a copy of chatData to chatHistory.
+     * @param {boolean} toast toast that the chat has saved.
+     */
+    async saveChatSnapshot(toast){
```
`showToast`

DeclineThyself replied:
Renamed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/ad55d4d919ad0f2f9822e46ce9f2fc37bee6cdc0

Cohee1207 commented on public/scripts/extensions/ChatUndoHistory/index.js:
```diff
@@ -0,0 +1,189 @@
+import { chat, clearChat, event_types, eventSource, printMessages, saveChatDebounced } from '../../../script.js';
```
You can use absolute path for import for less ../../..
`/script.js`

DeclineThyself replied:
Thank you: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/8803772fc630b43f495824c5f9f8bbd81f4a4d08

Cohee1207 commented on public/scripts/extensions/ChatUndoHistory/ui.js:
```diff
@@ -0,0 +1,209 @@
+/**
+ * Creates a range input.
+ */
+class rangeInput {
```
Use `PascalCase` for naming classes (even if not exported)

DeclineThyself replied:
Renamed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/032984ab57cd610b0182158d24e8e12df8a2bb0c

Commit:
Absolute imports.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2578971196

Commit:
showToast https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2578970342

Commit:
Case https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2578972403

Commit:
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2578969290

Commit:
Don't snapshot MESSAGE_UPDATED.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2578969290

Commit:
Renamed to `undo`.

Cohee1207 commented on public/scripts/extensions/undo/settings.html:
```diff
@@ -0,0 +1,26 @@
+<div id="undo_container" class="extension_container">
```
Extension containers are meant to be put in the index HTML DOM
They are meant to provide a static place for the extension settings, so that they won't change order on every page load.

DeclineThyself replied:
Thanks for explaining, Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/7d5082d79283c4eae2a14c49e00d86468eabfcf5

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Fixed  bug, applyDiff alters the lhs object.
Renamed var.

Commit:
Moved undo_container.

Cohee1207 commented:
I need some time to get to testing this. Please give me at least a couple of days.

Commit:
Fixed toggles.
EventListener != EventTarget

Cohee1207 committed:
Extract inline styles into a CSS file

Cohee1207 committed:
Merge branch 'staging' into feat/undo-history

Cohee1207 commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,215 @@
+ * Creates a range input.
+ */
+class RangeInput {
+    constructor( id, title, { dataStore = extension_settings[extensionName], callback = (id, value) => {}, category = extensionName, min = 0, max = 10000, step = 100, defaultValue = 1000, runCallbackOnLoad = true } = {}) {
```
It has a `step` of 100 but the default values are 20 and 512

DeclineThyself replied:
Done: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/b83c090a53da3d27c1f2279fff65ab2fb38bc2af

Cohee1207 commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,215 @@
+    const saveVisibility = (_, value) => $('#undo_save_options').toggle(!value);
+
+    //Setting that toggles menuVisibility and saveVisibility.
+    const toggleMenuElement = new ToggleInput('show_menu_buttons', 'Hide the Undo/Redo Buttons from the Options Menu.', { defaultValue: false, callback: menuVisibility }).create();
```
I think these toggles should be inverted: "Show" instead of "Hide". Also we don't actually call the hamburger menu "options" anywhere except the code. Maybe cut the text to something like: `Show Undo/Redo in ☰`

DeclineThyself replied:
Done: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/a1f4e18e23d244fc13ef3b8ac12e3ea745a22c1f

Cohee1207 commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,215 @@
+        else { document.removeEventListener('keydown', processUndoHotkey); }
+    };
+
+    const toggleUndoHotkeyElement = new ToggleInput('toggle_ctrl_z', 'Enable the ctrl-z/ctrl-Z hotkeys.', { defaultValue: false, callback: toggleUndoHotkey }).create();
```
There should be a killswitch that disables the entire extension. Right now it only controls the visibility of the buttons and hotkeys, but looks like it will start recording chat histories right after installing, without any user consent.

DeclineThyself replied:
It can be toggled in the `Manage Extensions` > `Built-in Extension` menu.
I'll add a redundant toggle.

DeclineThyself replied:
I added the toggle: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/25995a9ebcb74da92db2910b5ce6f2fae9f8d880
Should the extension be enabled or disabled by default?

Cohee1207 commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,215 @@
+
+    //Resetting the chatHistory is necassary to update chunk_size.
+    const resetDebounced = debounce(() => chatHistory.resetChatSnapshots(true), debounce_timeout.short);
+    const chunkSizeElement = new RangeInput('chunk_size', 'History Chunk Size. ⚠️ This will ERASE your history! Higher will use more memory, lower will reduce performance.', { min: 1, max: 10000, step: 10, defaultValue: defaultChunkSize, callback: () => resetDebounced(), runCallbackOnLoad: false }).create();
```
What actually is a "chunk" in this context? It's hard to understand by just looking at the settings panel. Is chunk a depth of the undo? Can we provide some guidance to user on how to determine an optimal size?

DeclineThyself replied:
I'm considering removing "chunking".
Currently, if `Chunk Size`/`fullHistoryInterval` is 3, then `chatHistory` would look like this:
```javascript
chatHistory = [
1fullChat,
2diff,
3diff,
4fullChat,
5diff,
6diff,]
```

The this allows for any chat to be loaded quickly by applying the diff to `fullChat`, and it's ~`1`/`fullHistoryInterval` the memory usage of storing full snapshots.

Instead, diffs can be stored of the next chat. So when loading `4fullChat`, `3diff` would be replaced with `3fullChat` by applying the diff forwards before removing `4fullChat`. 
```javascript
chatHistory = [
1diff,
2diff
3diff,
4fullChat,]
```

Then, if a snapshot is saved, first a `structuredClone` of `chat` would be added as `5fullChat` and `4fullChat` would be replaced with the diff of itself and `5fullChat`.

I didn't do this before because to load `1diff`, each diff subsequent would need to be applied going backwards from the `fullChat`.

It is simpler, and I don't believe the extension is likely to be modified in a way that allows users to load a snapshot that's few thousand items deep in `chatHistory`.

#### TLDR:
Do you prefer lower memory use, simpler code and less to explain to users, or future flexibility?

Cohee1207 replied:
I don't think you need to do structuredClones even once on loading a chat? What I would expect to see is a stack of diffs from the initial state when the chat was first loaded and an index of the last made diff. That way you can keep it the most memory efficient and avoid full clones.

DeclineThyself replied:
> I don't think you need to do structuredClones even once on loading a chat? What I would expect to see is a stack of diffs from the initial state when the chat was first loaded and an index of the last made diff. That way you can keep it the most memory efficient and avoid full clones.

If I were to stack diffs from the initial state (Which would need to be a `structuredClone`), the time to write a diff would scale at `O(n)`, where `n` is the distance from the `fullChat`.
Every previous diff would need to based upon the previous `fullChat`.
To get the previous `fullChat`, I would to apply the previous chat's diff with It's `fullChat` recursively until the the previous `fullChat` was initial state.

Here's my implementation. It's mostly untested but I would appreciate any advice.
Please skim my comments, I believe my above example was confusing and badly written.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/c7b5d7207b3b009d628dda86080a227c7c7e5916

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,189 @@
+        const max_length = extension_settings[extensionName]?.max_length ?? defaultMaxChatLength;
+
+        //Enforce the maximum chat length.
+        if (Array.isArray(this.chatData) && this.chatData.length >= max_length) {
```
Can we just remove this restriction? What is the reason for this to be limited?

DeclineThyself replied:
If there are users that have chats 10k+ messages long, I believe they should have a way to stop recording chat snapshots on every event.
Toggling the extension on and off between chats may be inconvenient.

Cohee1207 commented on public/scripts/extensions/undo/settings.html:
```diff
@@ -0,0 +1,30 @@
+        <div class="inline-drawer" id="undo-advanced-drawer">
+            <div class="inline-drawer-toggle inline-drawer-header">
+                <b data-i18n="Toggle snapshot events (Advanced)">
+                    Toggle snapshot events (Advanced)
```
Do we plan to leave this menu in the final version? I find it too technical even for the most powerful of power users. We usually don't provide control over debounce durations, etc. on the UI. It's better to agree on some "sane" and "good enough" defaults and strip the configuration to the essentials only.

DeclineThyself replied:
If `Snapshot Debounce Duration` and `History Chunk Size` are removed, can the rest stay?

Cohee1207 replied:
I think yes, at least for now.

DeclineThyself replied:
Done:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/c7b5d7207b3b009d628dda86080a227c7c7e5916
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/89863e478aa9feb1df474640d3bd4504e921f970

Cohee1207 commented:
Maybe remove toasts on undo? They stack up quickly when you do actions it in a quick succession and the UI updates themselves should provide enough of a visual feedback. And maybe disable the "Undo" when there's nothing to rollback to instead of toasting the error? It's really a normal state that there's nothing to come back to, not an error.

Cohee1207 commented on public/scripts/extensions/undo/manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "display_name": "Chat Undo History (ctrl + z)",
```
Let's finalize the keybind syntax to use on the UI (with this capitalization):
`Ctrl+Z` - undo
`Ctrl+Shift+Z` - redo
So there won't be any ambiguity in what actually needs to be pressed.

Cohee1207 replied:
Minor point: but Mac uses might need a bit different keybind indication:
`⌘+Z`
`⌘+⇧+Z`

DeclineThyself replied:
> Minor point: but Mac uses might need a bit different keybind indication: `⌘+Z` `⌘+⇧+Z`

I believe the `metaKey` is the `⌘` key.
https://github.com/SillyTavern/SillyTavern/pull/4819/files#diff-877e57394c51663a335521d0d11b1802e0fa2cafa928dc85a9f8e8c689b208a5R58

```javascript
async function processUndoHotkey(event) {
    if (!isInputElementInFocus()) {
        if ((event.ctrlKey || event.metaKey) && !event.altKey) {
            //Undo.
            event.key === 'z' && await chatHistory.loadPreviousSnapshot();
            //Redo.
            event.key === 'Z' && await chatHistory.loadNextSnapshot();
        }
    }
}
```

DeclineThyself replied:
Done: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/e402ff15b8d02b7bb12768b3553e5bbcba96ed3a

DeclineThyself commented:
> Maybe remove toasts on undo? They stack up quickly when you do actions it in a quick succession and the UI updates themselves should provide enough of a visual feedback. And maybe disable the "Undo" when there's nothing to rollback to instead of toasting the error? It's really a normal state that there's nothing to come back to, not an error.

Sometimes "undos" and appear to do nothing, I'd like to show which current of total (e.g., 3/10) "undos" is selected, and provide feedback if there's nothing to undo.
What would you suggest instead?

Is this okay?
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/190d67608b36a7ceb744d0cdedd5309a0316bb75

Commit:
Show Undo/Redo in ☰
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2590508900

Commit:
Changed defaults.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2590499464

Commit:
Ctrl+Shift+Z

Commit:
Fixed unused var.

Commit:
Added extension toggle, and it probably works this time.

Commit:
Prevent duplicate toasts.

DeclineThyself commented:
I found an upstream bug in `applyDiff`, the same mistake I've often made [elsewhere](https://www.github.com/SillyTavern/SillyTavern/pull/4828#issuecomment-3587455447).

It's caused [here](https://github.com/transformation-dev/blueprint/blob/4330d3045ef7b1838612e026bafaabd2637f55c6/packages/cloudflare-do-utils/src/apply-diff.js#L29), by deleting objects from an array while iterating forwards.
Instead, It should be spliced while iterating backwards.

```javascript
rhs = ["1", "2", "3"]
lhs = ["1"]
lhsDiff = diff(rhs, lhs)
restoredLHS = applyDiff(structuredClone(rhs), lhsDiff)
console.log("rhs", rhs)
console.log("lhs", lhs)
console.log("lhsDiff", lhsDiff)
console.log("restoredLHS", restoredLHS)

//This shouldn't return false.
lodash.isEqual(restoredLHS, lhs)
```

I created a PR upstream. Until it's fixed, I'll add the fixed function to the `undo` folder.

Commit:
Replaced applyDiff with fixed applyDelta.
https://github.com/transformation-dev/blueprint/blob/master/packages/cloudflare-do-utils/src/apply-delta.js

Commit:
Improve memory usage and remove chunking.

Commit:
Finished removing `deep-object-diff-apply`.

Commit:
Cleanup text.

Commit:
Removed user configurable `debounce_duration`.

Commit:
Fix toggle_extension.

Commit:
Refactor addSettings.

Commit:
Refactor UserInputs.

Commit:
Fixed bug.

Commit:
Fixed bug.
The diff of a message was not an object so `delta[key] instanceof Object` returned false.

Commit:
Don't store references in chatHistory.

DeclineThyself commented:
/gemini review
Gemini, focus on ChatHistory's implementation.
Is there another data structure that may be more memory efficient?

Commit:
Fixed possible recursion.
Removed `.prototype` check.
Added Disclaimer.

DeclineThyself commented on public/scripts/extensions/undo/applyDelta.js:
```diff
@@ -0,0 +1,28 @@
+// https://github.com/transformation-dev/blueprint/blob/master/packages/cloudflare-do-utils/src/apply-delta.js
+
+/* eslint-disable no-param-reassign */
+function innerApplyDelta(obj, delta) {
+    const keys = Object.keys(delta);
+    for (let i = keys.length - 1; i >= 0; i--) {
+        const key = keys[i];
+        if (Array.isArray(delta[key]) || delta[key] instanceof Set || delta[key] instanceof Map) {
+            obj[key] = delta[key];
+        } else if (delta[key] === undefined) {
+            if (Array.isArray(obj)) {
+                obj.splice(Number(key), 1);
+            } else {
+                delete obj[key];
+            }
+        } else if (typeof(delta[key]) === 'object') {
+            obj[key] = innerApplyDelta(obj[key] ?? {}, delta[key]);
+        } else {
+            obj[key] = delta[key];
+        }
+    }
+    return obj;
+}
```
> This function is vulnerable to prototype pollution. A specially crafted `delta` object (e.g., `JSON.parse('{"__proto__": {"polluted": true}}')`) could modify `Object.prototype`, leading to unexpected behavior or security issues across the application. You should add checks to prevent modification of `__proto__`, `constructor`, and `prototype`.

@Cohee1207 out of scope?
I believe only extensions could call this, they wouldn't gain any abilities they didn't already have.

> Additionally, the condition typeof(delta[key]) === 'object' is true for null, which can cause an error in the recursive call. You should also check for null to prevent this.

Fixed. https://github.com/SillyTavern/SillyTavern/pull/4819/commits/5ced0df96d6bdb40c7faa27c599e98e365a38733

Cohee1207 replied:
I don't have enough context to answer that. Could this potentially by triggered by data from imported chat files?

DeclineThyself replied:
> I don't have enough context to answer that. Could this potentially by triggered by data from imported chat files?

I don't know.
This is the first time I've heard of [prototype pollution](https://developer.mozilla.org/en-US/docs/Web/Security/Attacks/Prototype_pollution).

I believe that this shouldn't be an issue, but only in the way I'm using `applyDelta`.
I only call `applyDelta` with diffs of the `chat` object. `undo` never directly interacts with user input or any other distrusted input. (Lorebooks, Cards, Chat files, etc.)
I believe `chat` `message` would already need to have a polluted prototype to interact with `applyDelta`.

Regardless, I'll fix this before merging.

@gemini-code-assist, are my above assumptions correct?
Is this line enough to prevent prototype pollution?
`if (key === 'prototype' || key === '__proto__' || key === 'constructor') continue;`

DeclineThyself replied:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/0dbef50efe11f0e9aecbd74f619a1f32305d7703

DeclineThyself replied:
> Even if direct user input isn't involved, a compromised extension or an unforeseen interaction with other parts of the system could still expose this vulnerability.

@gemini-code-assist How?
Extensions should not access other extension's functions.
And a compromised extension could modify prototypes directly.

DeclineThyself replied:
@gemini-code-assist
Yes, I've already committed the fix, I'll test for prototype pollution elsewhere when I get the chance.

>Indirect Data Flow from Untrusted Sources: Even if applyDelta is currently only used with internal chat objects, the data within those chat objects might, at some point, originate from or be influenced by untrusted sources. For example, if a user imports a chat file (as @Cohee1207 alluded to), and that file contains a maliciously crafted object with a __proto__ key, and this object somehow makes its way into the delta parameter of applyDelta, the function could inadvertently pollute Object.prototype.

Please explain how my extension was vulnerable, I don't understand.
I am not parsing JSON anywhere in `undo` and `diffs` are only created off the `chat` object.
Wouldn't `chat` already need to have a polluted prototype to be stored in a diff, then enter `applyDelta`?

Please keep your responses more concise and directly answer my questions.
Others shouldn't need to read an essay to follow what's happened in your code review.

DeclineThyself replied:
@Cohee1207
TLDR:
Fixed https://github.com/SillyTavern/SillyTavern/commit/0dbef50efe11f0e9aecbd74f619a1f32305d7703.

DeclineThyself commented:
/gemini review
I'm asking for another review because Gemini ignored my code the first time.

DeclineThyself commented on public/scripts/extensions/undo/applyDelta.js:
```diff
@@ -0,0 +1,33 @@
+    return obj;
+}
+
+/**
+ * This function is only intended for use in ChatHistory until it gets fixed upstream.
+ * applyDelta has known issues, Don't use it.
+ * https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2595634539
+ */
+export function applyDelta(obj, delta) {
+    return innerApplyDelta(obj, delta);
+}
```
Yes, I added that comment.

`deep-diff` would not save memory. https://www.npmjs.com/package/deep-diff

DeclineThyself commented on public/scripts/utils.js:
```diff
@@ -2896,3 +2896,32 @@ export function createTimeout(ms, errorMessage = '') {
+ * @param {KeyboardEvent} event
+ * @returns {boolean}
+ */
+export function isModifiedKeyboardEvent(event) {
+    return (event instanceof KeyboardEvent &&
+        event.shiftKey ||
+        event.ctrlKey ||
+        event.altKey ||
+        event.metaKey);
+}
```
Out of scope, that was moved from `RossAscends-mods.js`.
I made a PR: https://github.com/SillyTavern/SillyTavern/pull/4866

Commit:
Fixed inconsistent `defaultMaxChatLength`.

DeclineThyself commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,245 @@
+    async loadChatSnapshot(offset) {
+        const t1 = performance.now();
+        const index = this.chatHistoryIndex + offset;
+        const maximumChatLength = extension_settings[extensionName]?.max_length ?? 512;
```
Fixed.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/94952cb39a1f3d6796620d4c885e95082ef9b75a

Commit:
Off by one in maxChatLength.

Commit:
Fixed: Edits were not saving.

DeclineThyself commented:
> /gemini review
> Gemini, focus on ChatHistory's implementation.
> Is there another data structure that may be more memory efficient?

@gemini-code-assist, Answer my above question.

DeclineThyself commented:
For any humans, this is incorrect:
> memory footprint of approximately "2*chat" (one full snapshot and one diff for the current state),

No, not one diff. `2*chat` includes one full snapshot, and as many diffs as there are messages.
One full snapshot, then all diffs added together should approximately equal the size of the full chat during normal usage.

It may be much more. For example, If an extension (Or the `chatTree`) were to add then remove many messages repeatedly, then the diffs would be much larger.

DeclineThyself commented:
@Cohee1207, Should I Re-request review once I believe I've resolved all the previous review's comments?

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Prevent Prototype pollution.

Cohee1207 committed:
Merge branch 'staging' into feat/undo-history

Cohee1207 commented on public/scripts/extensions/undo/buttons.html:
```diff
@@ -0,0 +1,25 @@
+        <span data-i18n="Redo">Redo</span>
+    </a>
+</div>
+<div id="undo_save_options">
```
What's the significance of save/discard buttons for a user? In what scenario they are meant to be used? Is it to perform manual checkpointing? Shouldn't every chat action that's triggered by an event make a save?

By the way, you can likely leave both of these if you find them useful, but under the extension settings, not this menu.

DeclineThyself replied:
> What's the significance of save/discard buttons for a user? In what scenario they are meant to be used? Is it to perform manual checkpointing? Shouldn't every chat action that's triggered by an event make a save?

Some users may disable all 'Snapshot Events' in the advanced menu.
This allows those users to manually create snapshots.

> By the way, you can likely leave both of these if you find them useful, but under the extension settings, not this menu.

By default, the `Show Save/Reset in ☰`  toggle is disabled, and the buttons are not visible in this menu.

May I leave it as-is?

Cohee1207 commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,235 @@
+        }
+    });
+
+    const disableExtension = new ToggleInput('toggle_extension', 'Disables the extension.', { defaultValue: false, callback: toggleExtension, runCallbackOnLoad: false }).create();
```
Extension kill switches are usually inverted. Web search for example:
<img width="158" height="61" alt="image" src="https://github.com/user-attachments/assets/75875fae-3ec6-4243-9046-ef3049b2d532" />

DeclineThyself replied:
Done:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/8e44f0155b286be304c2e6ff3fa33cbea6cad7ec

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,245 @@
+            if (newChat.length > oldChatLength) { await eventSource.emit(event_types.MESSAGE_RECEIVED, undefined, 'undo'); }
+            if (newChat.length < oldChatLength) { await eventSource.emit(event_types.MESSAGE_DELETED, undefined, 'undo'); }
+
+            toastr.clear(); //Remove the previous toast to prevent buildup.
```
This is not needed. You can launch a toast with preventDuplicates option

DeclineThyself replied:
I didn't know about `preventDuplicates`, Thank you.
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1f6429265c4e84f3736f5c280c2b0d26cf25b09a

I still mildly prefer `toastr.clear` because `preventDuplicates` gives no visual feedback when a button is clicked again.
`toastr.clear` shows that the toast has been replaced.
May I revert this commit?

Cohee1207 replied:
toastr.clear:
1. Is too sudden 
2. Clears other unrelated toasts

DeclineThyself replied:
Okay, thanks for explaining.

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,245 @@
+            if (newChat.length < oldChatLength) { await eventSource.emit(event_types.MESSAGE_DELETED, undefined, 'undo'); }
+
+            toastr.clear(); //Remove the previous toast to prevent buildup.
+            toastr.success(`Chat ${this.chatHistoryIndex + 1}/${this.chatHistory.length} has been loaded.`);
```
I don't think that the term 'chat' is appropriate since it implies a specific chat state or snapshot, not the 'chat' as top level entity.

DeclineThyself replied:
Agreed, I changed it to 'Snapshot'. https://github.com/SillyTavern/SillyTavern/pull/4819/commits/5788e6d406bca80e4d9bd3a122de31765d20d279

Cohee1207 commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,245 @@
+        }
+        else {
+            toastr.clear();
+            toastr.error(`Chat ${index + 1}/${this.chatHistory.length} does not exist!`);
```
Maybe make these toasts optional? I'd prefer them to be quiet but default tbh, but you and others may want to have the toasts.

DeclineThyself replied:
By default, I'd prefer to annoy users with a toast that can be disabled, instead of confusing them when the button appears to do nothing.

The toasts are optional now: https://github.com/SillyTavern/SillyTavern/pull/4819/commits/3a4df17d4ab94446ec7a75ba28a1cd566b9a3541

Cohee1207 committed:
Update display name

Commit:
Inverted extension killswitch:
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2607912458

Commit:
Typo.

Commit:
Translate toasts.

Commit:
Toastr preventDuplicates.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2607915377

Commit:
Use the term 'Snapshot' instead of 'Chat'.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2607917146

Commit:
Made undo/redo toasts optional.
https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2607919477

Cohee1207 commented:
Not sure if that's intended, but clicking the killswitch toggles all the checkboxes under the extension

DeclineThyself commented:
> Not sure if that's intended, but clicking the killswitch toggles all the checkboxes under the extension

Yes, that's intended.

Commit:
Renamed settings keys and backported NeoTavern changes.

Cohee1207 commented:
Why? I'm a bit confounded about this behavior. Killswitch should just toggle the extension's functionality on and off.

DeclineThyself commented:
> Why? I'm a bit confounded about this behavior. Killswitch should just toggle the extension's functionality on and off.

Calling `.click` on each button was simplest way to implement a kill switch, Event listeners and options buttons need to be removed. 
I prefer to visually show the user what has been toggled off.
If the extension is 'disabled' then the hotkeys will not trigger and the buttons are removed, so the `Show Undo/Redo` toggle shouldn't display a check mark.

Cohee1207 commented:
By killswitch I meant a setting that completely shuts the extension down. Since it's enabled by default and will consume memory even if unwanted. Disabling all events for that is not a convenient way.

Commit:
Clear chatHistory when the extension is disabled.

DeclineThyself commented:
> By killswitch I meant a setting that completely shuts the extension down. Since it's enabled by default and will consume memory even if unwanted. Disabling all events for that is not a convenient way.

I still don't understand why the killswitch is needed in addition to the toggle in `Manage Extensions`.

The current `Enable the extension` switch stops all future snapshots, disables all listeners and removes the buttons in ☰.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/ff81bdd1c6cb6440b62e13738da9677548999876 It will now also delete the current history to free memory.

Is this what you wanted?

Commit:
Typo.

Cohee1207 commented:
All built-in extensions come enabled by default. You would have to go there and disable it manually.
Can I maybe move this to a separate repo / make it downloadable via the assets list? That way I won't have any trepidations regarding its functionality being on by default.

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Refactored UI and disabled extension by default.

Commit:
Only reset snapshots if the extension is enabled.

DeclineThyself commented:
> That way I won't have any trepidations regarding its functionality being on by default.

I had thought this is an unobtrusive feature that most users would want by enabled default.
The extension now defaults to disabled and will not effect performance until enabled.
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/c431a089cc075780cba53fc28b95ef7ec3b77434


>  All built-in extensions come enabled by default. You would have to go there and disable it manually. Can I maybe move this to a separate repo / make it downloadable via the assets list? 

I would prefer to leave the extension visible, this feels like a basic feature that should be built-in to ST.
If it's not visible then most users will not find it. I typically don't look for an undo button until I've already deleted my work.

DeclineThyself commented:
> That way I won't have any trepidations regarding its functionality being on by default.

> The extension now defaults to disabled and will not effect performance until enabled.

@Cohee1207, Is this okay?

Cohee1207 commented:
@Wolfsblvt please review, need your opinion on built-in vs downloadable

Wolfsblvt commented:
> @Wolfsblvt please review, need your opinion on built-in vs downloadable

I think this is similar to the Input history on the Chat input bar. It's a nice-to-have. But considering the memory footprint of this, and the complexity this needs (700 lines of code), I am a bit hesitant too.

The biggest difference between having it as an assets list extension with recommendation, and a built-in extension is also *ownership*. In the sense of, any core extension that is built-in needs to be maintained by the dev team for anything that will happen in the future. I don't feel like the code of this is too complicated to not be maintainable, so that's not a direct ko criterium, but its something to think about for sure.

As we've already talked about on Discord, I think this is more of a presentation issue of recommended extensions that this needing to be built-in. I agree that being on the asset list is really not a good argument to "people can easily find it" right now.

Wolfsblvt commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,248 @@
+        const offsetChatDiff = this.chatHistory[offsetId];
+        if (typeof(offsetChatDiff) == 'object') {
+            return applyDelta(structuredClone(fullChat), structuredClone(offsetChatDiff));
+        }
```
How often do you expect the objects in this array you build yourself to not be of type `object`?
And more important questions, if they aren't, what do you return? Right now, it's implicitly `undefined`, but not documented in this function.

DeclineThyself replied:
> How often do you expect the objects in this array you build yourself to not be of type object?
When they're undefined.

>  And more important questions, if they aren't, what do you return? Right now, it's implicitly undefined, but not documented in this function.
Thank you, fixed:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1a771dd2b2dbd309759c8a3e4532edd2df0a542a

Wolfsblvt commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,248 @@
+
+        //Create the offsetChat from by updating the fullChat with the diff.
+        const offsetChat = this.adjacentChat(offset, fullChat);
+        if (typeof(offsetChat) == 'object') {
```
Some here, why checking if its object? It should always be an object. Or, if adjacentChat can actually not be there, you should check for null/undefined, and make the jsdoc on the function explicit.

DeclineThyself replied:
> Some here, why checking if its object? It should always be an object. Or, if adjacentChat can actually not be there, you should check for null/undefined, and make the jsdoc on the function explicit.

Thank you.
I fixed the JSDoc and improved the comments:
https://github.com/SillyTavern/SillyTavern/pull/4819/commits/1a771dd2b2dbd309759c8a3e4532edd2df0a542a

Wolfsblvt commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,248 @@
+        this.chatHistoryIndex = 0;
+
+        this.chatHistory = [structuredClone(this.chatData)];
+        if (showToast) toastr.warning(t`You now have ${this.chatHistory.length} snapshots.`, t`Success.`, { preventDuplicates: true });
```
Why the preventDuplicates? Shouldn't this always toast when you warn?
Also, this might not work like you expect it to. It will prevent duplicates with the exact same `this.chatHistory.length`. If its different, it'll make another toast, that then can't appear.

DeclineThyself replied:
@Wolfsblvt, I agree.
@Cohee1207 does not. Please discuss this with Cohee and I will implement whatever is decided.
I still prefer the tradeoffs of toastr.clear().
Should I add a `Prevent toasts from Stacking up` dropdown with options like `preventDuplicates`, `Clear` and `Disabled` .

Context:

https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3614476951
> Maybe remove toasts on undo? They stack up quickly when you do actions it in a quick succession and the UI updates themselves should provide enough of a visual feedback. And maybe disable the "Undo" when there's nothing to rollback to instead of toasting the error? It's really a normal state that there's nothing to come back to, not an error.

https://github.com/SillyTavern/SillyTavern/commit/190d67608b36a7ceb744d0cdedd5309a0316bb75
I added `toastr.clear` to prevent them from stacking up.

https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2608215239
>  toastr.clear:
    Is too sudden
    Clears other unrelated toasts

Cohee didn't like the problems that come with `toastr.clear`, I swapped to `preventDuplicates` as requested.
https://github.com/SillyTavern/SillyTavern/commit/1f6429265c4e84f3736f5c280c2b0d26cf25b09a


https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2607919477
> Maybe make these toasts optional? I'd prefer them to be quiet but default tbh, but you and others may want to have the toasts.

I made the toasts optional. https://github.com/SillyTavern/SillyTavern/commit/3a4df17d4ab94446ec7a75ba28a1cd566b9a3541

Cohee1207 replied:
You can keep a reference to the latest toast and remove it from DOM instead of calling toastr.clear that clears ALL (even unrelated) toasts.

Wolfsblvt replied:
Keeping a "stack" of toasts that are undo/redo actions with when you send them, and clear all references that are still in the timeout timeframe sounds a lot better. Or maybe, just show the last two or three of those?
And you should remember only toasts coming from this extension.

I can see how `clear()` looks really bad if you simply call that one. But making the toasts be a "fire once and then never" is just... plain wrong. It does not show you what you are actually doing. It does not confirm chat history actions that you are doing.
There is likely a more reasonable system to implement this and still prevent toast spam.

Wolfsblvt commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,248 @@
+
+        //Enforce the maximum chat length.
+        if (Array.isArray(this.chatData) && this.chatData.length > maxChatLength) {
+            if (showToast) toastr.error(t`It's in 'Extensions > Chat Undo History > Max chat length'`, t`You cannot save the chat because it's ${this.chatData.length - maxChatLength} messages longer than your max chat length limit (${maxChatLength}). (Check Settings.)`, { preventDuplicates: true });
```
Especially this toast shouldn't be `preventDuplicates` for sure, shouldn't it?

Also, careful with toastr message designs. Second argument is "title", not the message. And while we often have the title be a sentence, I am not a big fan of it. Especially if its long. It just looks weird.
Title should be the short hook informing you what this toast is about. Then, the text can explain everything. You can use multiline toasts if you need to explain more. Worst case, use a different longer timeout.

DeclineThyself replied:
> Especially this toast shouldn't be preventDuplicates for sure, shouldn't it?

Please see: https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2637928553

> Also, careful with toastr message designs. Second argument is "title", not the message. And while we often have the title be a sentence, I am not a big fan of it. Especially if its long. It just looks weird.
> Title should be the short hook informing you what this toast is about. Then, the text can explain everything. You can use multiline toasts if you need to explain more. Worst case, use a different longer timeout.

Thank you, I'll correct the titles.

Wolfsblvt commented on public/scripts/extensions/undo/index.js:
```diff
@@ -0,0 +1,248 @@
+
+        //Max history cannot be less than zero.
+        if (0 >= maxUndoSnapshots) {
+            if (showToast) toastr.error(t`It's in 'Extensions > Chat Undo History > Max Undo Snapshots'`, t`You cannot save the chat because your Max Undo Snapshots is set to ${maxUndoSnapshots}. (Check Settings.)`, { preventDuplicates: true });
```
Same here, title a bit too long.
(Not gonna comment on all following for that, there are several more.)

Wolfsblvt commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,271 @@
+    };
+
+    //When a setting is toggled, the enableExtension button may need to be enabled or disabled.
+    eventSource.on(`extension_${extensionName}`, (setting, value) => {
+        if ((['showMenuButtons', 'showSaveButtons', 'enableCtrlZ', 'showToasts'].includes(setting) || setting.includes('message_'))) {
+            //When any setting is toggled, show the extension as enabled.
+            if (value) {
+                enableExtension.set(value);
+            }
+            //Or, If all settings are disabled, show the extension as disabled.
+            else if (!(toggles.map((toggle) => toggle.toggled())).some((enabled) => { return enabled; })) {
+                enableExtension.set(value);
+            }
+        }
+    });
```
Is that the only built-in extension that does something like that? I am not sure if auto-toggling the extension off is like.. a good idea. That breaks the usual system of enabling/disabling extensions manually as users choice.

DeclineThyself replied:
> Is that the only built-in extension that does something like that?

I believe so.

> I am not sure if auto-toggling the extension off is like.. a good idea. That breaks the usual system of enabling/disabling extensions manually as users choice.

This is not the toggle in `Manage Extensions`.
This is a separate toggle I added in response to Cohee's request for a killswitch. https://github.com/SillyTavern/SillyTavern/commit/25995a9ebcb74da92db2910b5ce6f2fae9f8d880


https://github.com/SillyTavern/SillyTavern/pull/4819#discussion_r2590513674
> There should be a killswitch that disables the entire extension. Right now it only controls the visibility of the buttons and hotkeys, but looks like it will start recording chat histories right after installing, without any user consent.



https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3648525508

> > By killswitch I meant a setting that completely shuts the extension down. Since it's enabled by default and will consume memory even if unwanted. Disabling all events for that is not a convenient way.
> 
> I still don't understand why the killswitch is needed in addition to the toggle in `Manage Extensions`.
> 
> The current `Enable the extension` switch stops all future snapshots, disables all listeners and removes the buttons in ☰. [ff81bdd](https://github.com/SillyTavern/SillyTavern/commit/ff81bdd1c6cb6440b62e13738da9677548999876) It will now also delete the current history to free memory.
> 
> Is this what you wanted?

The implementation is complex because I wanted to improve the UX.
If all toggles are disabled, then the `Enable Extension` toggle will show as disabled.
If the extension is disabled and any other toggle is enabled, the `Enable Extension` toggle shows as enabled.

The opposite would be inconsistent and confusing.

Personally, I would prefer to remove the toggle/killswitch entirely and trust that the user can find the toggle in `Manage Extensions`.
The memory usage of this extension is extremely optimized and should not be an issue.

Cohee1207 replied:
That's what I meant by a killswitch. Lots of other extensions do it like that.

```
if (!enabled) {
  return;
}
```

Wolfsblvt replied:
Okay, I misunderstood that. I still think auto-toggling a toggle you didn't click to "off" is a bit weird, but it doesn't hurt much.
If you disable all options that the extension provides, it can also turn off and tell you that the whole extension functionality is disabled. Yeah okay.

Wolfsblvt commented on public/scripts/extensions/undo/ui.js:
```diff
@@ -0,0 +1,271 @@
+    await addSettingsAdvancedToggles(toggles);
+}
+
+/**
+ * Boilerplate.
+ */
+class UserInput {
+    constructor( id, title, { dataStore = extension_settings[extensionName], callback = (id, value) => {}, category = extensionName, runCallbackOnLoad = true, parent = undefined } = {}) {
+        this.id = id;
+        this.title = title;
+        this.dataStore = dataStore;
+        this.callback = callback;
+        this.category = category;
+        this.runCallbackOnLoad = runCallbackOnLoad;
+        this.parent = parent;
+        this.element = undefined;
+    }
+}
```
While I do think that having common UI input elements for both extensions and also our main user settings and all, I am not sure if this specific extension is the right place to start with those?

I mean sure, we have to start *somewhere*. But if you make this here, it'll be a lot harder to refactor it out in the future, and standardize it for other extensions, etc. Or, it'll just stay here, and any other extension or future core code will just duplicate it.

Generally, I think this adds what the code comment already says. A lot of boilerplate for this extension for... minimal gain, for having like under 10 toggles, that could easily be implemented with HTML definitions and a few click/toggle handlers like in other places.

I do like the idea. Just questioning it being part of this PR/extension.

DeclineThyself replied:
> While I do think that having common UI input elements for both extensions and also our main user settings and all, I am not sure if this specific extension is the right place to start with those?

> I mean sure, we have to start somewhere. But if you make this here, it'll be a lot harder to refactor it out in the future, and standardize it for other extensions, etc. Or, it'll just stay here, and any other extension or future core code will just duplicate it.

I designed these functions to be moveable with minimal changes. A refactor will not require much than a copy paste and import update.

> I do like the idea. Just questioning it being part of this PR/extension.

I agree.
In a new PR, May I move these functions to utils and replace most of the hardcoded HTML `power-user.js` toggles and sliders with `ToggleInput` and `RangeInput`?
I don't know if there will  be a performance impact compared to hardcoded HTML.

Wolfsblvt replied:
I don't mind this. I think it makes sense, yeah. Both for power user toggles, and also all the toggles from built-in extensions we already have.

If @Cohee1207 doesn't mind, we can keep this in this PR here, for now, until it gets refactored.
Just wanted to say that I don't think this scopes well with this PR specifically. Not that I don't like this implementation.

Wolfsblvt reviewed (commented):
I had a few remarks when going through the code.

Mostly minor stuff, about the toastr design (no good definitions for toast headers, and needless `preventDuplicates` for nearly all toasts),
as well as a nitpick with the estension settings UI boilerplate.

Generally tested this, and the functionality seems to be working. I agree this is a cool feature to have and there are likely people who'd use this a lot.

Just same as Cohee, I am not sure about having it as a built-in extension but disabled by default is a good choice.

DeclineThyself commented:
> The biggest difference between having it as an assets list extension with recommendation, and a built-in extension is also _ownership_. In the sense of, any core extension that is built-in needs to be maintained by the dev team for anything that will happen in the future. I don't feel like the code of this is too complicated to not be maintainable, so that's not a direct ko criterium, but its something to think about for sure.

Thanks for explaining, I appreciate it.

> Just same as Cohee, I am not sure about having it as a built-in extension but disabled by default is a good choice.

> As we've already talked about on Discord...

I haven't spoken to you on discord. Are you referring to Cohee?

Cohee1207 commented:
> I haven't spoken to you on discord. Are you referring to Cohee?

Yes, we had a little chat on the onboarding flow with enabling / downloading extensions on first run.

Any big objections against moving this to a downloadable extension?

Commit:
Merge remote-tracking branch 'origin/staging' into feat/undo-history

Commit:
Check if undefined.
Fixed JSDoc, improved comments.

DeclineThyself commented:
> Yes, we had a little chat on the onboarding flow with enabling / downloading extensions on first run.

I made an account a while ago, I was unable to speak due to the week long waiting period.
I'll check if that's passed.

> Any big objections against moving this to a downloadable extension?

No, I appreciate your consideration, Thank you.
In the future, If possible please let me know initially if something isn't fit for this repository. I wouldn't have spent the effort knowing that my work will impact a fraction the users.

How do I migrate this to be a downloadable extension?

Cohee1207 commented:
> In the future, If possible please let me know initially if something isn't fit for this repository. I wouldn't have spent the effort knowing that this my work will impact a fraction the users.

That's usually my first question for any substantial changes proposed to core. [Perhaps I wasn't insistent enough?](https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3583130884)

> How do I migrate this to be a downloadable extension?

Just move all files to a separate repository and adjust import paths (all relative paths should go one directory down). To make it easier, change the imports to use absolute paths, e.g. `/script.js`

Then send a PR to Content repo (https://github.com/SillyTavern/SillyTavern-Content). See requirements in the README.

DeclineThyself commented:
> [Perhaps I wasn't insistent enough?](https://github.com/SillyTavern/SillyTavern/pull/4819#issuecomment-3583130884)
Let's move into an extension. If you don't want to make a downloadable one, start with a built-in.

I misunderstood. I had thought you confirmed built-in was acceptable.

Wolfsblvt commented:
> I misunderstood. I had thought you confirmed built-in was acceptable.

I mean, it's acceptable, generally. It still has to be justified, based on the trade-offs. I think that's the big point here. Actual use-case for most of the userbase vs. need to be maintained, code-ownership, etc etc.

DeclineThyself commented:
> Then send a PR to Content repo (https://github.com/SillyTavern/SillyTavern-Content). See requirements in the README.

>> Extensions must be compatible with the latest release version of SillyTavern. Please be ready to update your extension if something in the core changes.

I'm unable to meet the requirements.

I thank both of you for your time reviewing my code. I appreciate the chance to learn and improve.
